spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -139,39 +146,63 @@ public Subscriber subscribe(String subscription,
 
-		return receivedMessageList.size() > 0 ?	receivedMessageList.get(0) : null;
+		return receivedMessageList.size() > 0 ? receivedMessageList.get(0) : null;
 	}
 
 	public SubscriberFactory getSubscriberFactory() {
 		return this.subscriberFactory;
 	}
 
 	@Override
-	public void ack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
+	public ListenableFuture<Collection<String>> ack(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
 		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be null.");
 
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages).forEach(this::ack);
+		Map<String, List<String>> acknowledgeablePubsubMessagesMap =
+				groupAcknowledgeableMessages(acknowledgeablePubsubMessages);
+
+		Collection<String> ackIds = new LinkedList<>();
+		for (Map.Entry<String, List<String>> entry : acknowledgeablePubsubMessagesMap.entrySet()) {
+			ack(entry.getKey(), entry.getValue());
+			((LinkedList<String>) ackIds).addAll(entry.getValue());
+		}
+
+		SettableListenableFuture<Collection<String>> settableListenableFuture = new SettableListenableFuture<>();
+		settableListenableFuture.set(ackIds);
+
+		return settableListenableFuture;
 	}
 
 	@Override
-	public void nack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
-		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be null.");
-
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages).forEach(this::nack);
+	public ListenableFuture<Collection<String>> nack(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
+		return modifyAckDeadline(acknowledgeablePubsubMessages, 0);
 	}
 
 	@Override
-	public void modifyAckDeadline(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages,
-			int ackDeadlineSeconds) {
+	public ListenableFuture<Collection<String>> modifyAckDeadline(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages, int ackDeadlineSeconds) {
 		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be null.");
 		Assert.isTrue(ackDeadlineSeconds >= 0, "The ackDeadlineSeconds must not be negative.");
 
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages)
-				.forEach((sub, ackIds) -> modifyAckDeadline(sub, ackIds, ackDeadlineSeconds));
+		Map<String, List<String>> acknowledgeablePubsubMessagesMap =
+				groupAcknowledgeableMessages(acknowledgeablePubsubMessages);
+
+		Collection<String> subscriptionNames = new LinkedList<>();
+		for (Map.Entry<String, List<String>> entry : acknowledgeablePubsubMessagesMap.entrySet()) {
+			modifyAckDeadline(entry.getKey(), entry.getValue(), ackDeadlineSeconds);
+			subscriptionNames.add(entry.getKey());
+		}
+
+		SettableListenableFuture<Collection<String>> settableListenableFuture = new SettableListenableFuture<>();
+		settableListenableFuture.set(subscriptionNames);
```

Comment:
This doesn't work. The `SettableListenableFuture` you're returning is completely unhinged from the actual calls to `modifyAckDeadline`.

Reply:
@meltsufin Agreed ... I'll probably more direction from you and @artembilan. The whole `ListenableFuture` use seems to be baffling to me for the receive case.

Reply:
@meltsufin Would something like this work?

```
	@Override
	public ListenableFuture<Collection<Result>> modifyAckDeadline(
			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages, int ackDeadlineSeconds) {
		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be null.");
		Assert.isTrue(ackDeadlineSeconds >= 0, "The ackDeadlineSeconds must not be negative.");
		
		Collection<Result> results = new LinkedList<>();
		
		SettableListenableFuture<Collection<Result>> settableListenableFuture = new SettableListenableFuture<>();
		settableListenableFuture.set(results);
		
		Map<String, List<String>> acknowledgeablePubsubMessagesMap =
				groupAcknowledgeableMessages(acknowledgeablePubsubMessages);
		
		for (Map.Entry<String, List<String>> entry : acknowledgeablePubsubMessagesMap.entrySet()) {
			Result result = new Result(entry.getKey(), entry.getValue());
			
			ApiFuture<Empty> apiFuture = modifyAckDeadline(entry.getKey(), entry.getValue(), ackDeadlineSeconds);
			ApiFutures.addCallback(apiFuture, new ApiFutureCallback<Empty>() {
				
				@Override
				public void onFailure(Throwable throwable) {
					result.setThrowabe(throwable);
				}

				@Override
				public void onSuccess(Empty empty) {

				}
			});
			
			results.add(result);
		}

		return settableListenableFuture;
	}
```

... where Result is ...

```	
	class Result {

		private String subscription;

		private Collection<String> ackIds;

		private Throwable throwable;
		
		public Result(String subscription, Collection<String> ackIds) {
			this.subscription = subscription;
			this.ackIds = ackIds;
		}
		
		public void setThrowabe(Throwable throwable) {
			this.throwable = throwable;
		}
		
		public String getSubscription() {
			return this.subscription;
		}
		
		public Collection getAckIds() {
			return this.ackIds;
		}
		
		public Throwable getThrowable() {
			return this.throwable;
		}
	};
```